### PR TITLE
Document that bcrypt doesn't handle long passwords

### PR DESCRIPTION
**Pull request by alex**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -84,6 +84,21 @@ Another one of bcrypt's features is an adjustable prefix to let you define what
...
+
+The bcrypt algorithm only handles passwords up to 72 characters, any characters
+beyond that are ignored. To work around this, a common approach is to hash a
+password with a cryptographic hash, such as ``sha512`` before hasing it with
+``bcrypt``:
+
+.. code:: pycon
+
+    >>> password = b"an incredibly long password" * 10
+    >>> hashed = bcrypt.hashpwd(
```
Comment (dstufft) [inline README.rst]: It's ``bcrypt.hashpw``.

**Comment by dstufft** [comment]
Would it be worthwhile to have a helper function over this? While I don't think it's particularly hard to do (and the documentation is good either way) would something like:

```python
# In bcrypt/__init__.py
import hashlib

def betterhashpw(password, salt):  # I can't think of a good name...
    hashed_password = hashlib.sha512(password).digest()
    return hashpw(hashed_password, salt)
```

Personally, I'm kinda ehhh on it, because by the time you get around to dealing with "well what happens if sha512 needs to be replaced at some point" you start needing to solve migrating old password and being able to bake something into the hash identifier so you know which hash func needs to be used.

**Comment by alex** [comment]
Should we just throw an exception for `len(password) > 72`?

On Sat, Oct 17, 2015 at 12:56 PM, Donald Stufft <notifications@github.com>
wrote:

> Would it be worthwhile to have a helper function over this? While I don't
> think it's particularly hard to do (and the documentation is good either
> way) would something like:
>
> # In bcrypt/__init__.pyimport hashlib
> def betterhashpw(password, salt):  # I can't think of a good name...
>     hashed_password = hashlib.sha512(password).digest()
>     return hashpw(hashed_password, salt)
>
> Personally, I'm kinda ehhh on it, because by the time you get around to
> dealing with "well what happens if sha512 needs to be replaced at some
> point" you start needing to solve migrating old password and being able to
> bake something into the hash identifier so you know which hash func needs
> to be used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/bcrypt/pull/52#issuecomment-148933930>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by dstufft** [comment]
That would be a breaking change that I'm not sure is worth it. Perhaps some sort of warning would be more appropriate? As far as I know there aren't any experts who consider the 72 character a security problem and they all treat wrapping the password with a cryptographic hash as security hardening rather than fixing a problem.

**Comment by dstufft** [comment]
Of course, both the warning and the exception have the problem that it's unlikely to fail during development and more likely to happen during production, since most developers are unlikely to be using > 72 character passwords as tests.

**Comment by dstufft** [comment]
Another thought: We should probably add some documentation (in another PR) that directs people not to use this library directly and instead use it as part of passlib or Django's hasher framework or something. This library is more along the lines of cryptography's hazmat layer then the recipes layer.

**Comment by alex** [comment]
Tests are failing becuase coverage dorpped 3.2, shall we drop 3.2 as well?

**Comment by dstufft** [comment]
Seems reasonable to me.

**Pull request merged**